### PR TITLE
Dangling item id in partial specializations

### PR DESCRIPTION
**Pull request by fitzgen**
See individual commit messages for details.

r? @emilio

### CODE REVIEW

**Comment by emilio** [comment]
Looks good to me, just one question. If you could add a test for this, that'd be neat.

r=me, ideally with test (I'd look into specifying something like `--whitelist-type dummy` so we need to scan the items or something)

File: src/ir/context.rs [hunk 1]
```diff
@@ -505,9 +505,15 @@ impl<'ctx> BindgenContext<'ctx> {
                 found_invalid_template_ref = true;
             }
             if c.kind() == CXCursor_TypeRef {
+                // The `with_id` id will potentially end up unused if we find we
+                // already have an item (with its own distinct `ItemId`) for
+                // this type, so if we pass `with_id` as the parent, it is
+                // potentially a dangling reference. Instead, use the canonical
+                // template declaration as the parent. It is already parsed and
+                // has a known-resolvable `ItemId`.
```
Comment (emilio) [inline src/ir/context.rs]: I think this comment is not as accurate as it should. This ends up unused if we peek the early return in line 555 (the one with the `error!` thing, it's not about finding other type that is the same.

**Comment by fitzgen** [comment]
@bors-servo r=emilio

I tried creating a test case, but couldn't build one that triggered this behavior :-/

**Pull request merged**